### PR TITLE
For review

### PR DESCRIPTION
**Pull request by OmelchenkoE**

### CODE REVIEW

File: src/main/SimpleListImpl.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
+package main;
+
+public class SimpleListImpl implements SimpleList {
```
Comment (drykiss0) [inline src/main/SimpleListImpl.java:5]: **General Implementation comment**
You are re-copying array every time you add/remove an element in order to expand/shrink array size. This is extremely ineffective. Please implement approach - where you define a constant MIN_ARRAY_SIZE (let's say with a value of 5). You should create initial array with this size. You will initially have space for 5 elements. If you need to add a 6th element - grow your array by another MIN_ARRAY_SIZE - so you will have an array with space for 10 elements. For removal - do a similar thing- shrink array if there are at least MIN_ARRAY_SIZE empty places.

File: src/main/SimpleListImpl.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    private Object[] arr = new Object[0];
+
+    @Override
+    public Object get(int index) {
+        if (index < arr.length) {
```
Comment (drykiss0) [inline src/main/SimpleListImpl.java]: Please handle situation when index < 0

File: src/main/SimpleListImpl.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+            return arr[index];
+        } else {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
+    @Override
+    public Object set(int index, Object element) {
+        if (index >= arr.length) {
```
Comment (drykiss0) [inline src/main/SimpleListImpl.java]: Please handle situation when index < 0

File: src/main/SimpleListImpl.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+            throw new IndexOutOfBoundsException();
+        } else {
+            arr[index] = element;
+            return arr[index];
```
Comment (drykiss0) [inline src/main/SimpleListImpl.java]: The returned result is wrong. Please take a look at java.util.List javadoc of this method. You should return an element that was previously on given position.

File: src/main/SimpleListImpl.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+        }
+    }
+
+    @Override
+    public void add(int index, Object element) {
+        if (index >arr.length) {
```
Comment (drykiss0) [inline src/main/SimpleListImpl.java]: Please handle situation when index < 0

File: src/main/SimpleListImpl.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+    @Override
+    public Object remove(int index) {
+        if (index < size()) {
+            Object[] temp = arr;
+            arr = new Object[arr.length - 1];
+            System.arraycopy(temp, 0, arr, 0, index);
+            int aftterI = temp.length - index - 1;
+            System.arraycopy(temp, index + 1, arr, index, aftterI);
+            return true;
```
Comment (drykiss0) [inline src/main/SimpleListImpl.java]: Returned result is wrong. Please revisit javadoc of this method.

File: src/main/SimpleListImpl.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        } else {
+            throw new UnsupportedOperationException();
```
Comment (drykiss0) [inline src/main/SimpleListImpl.java]: Should be different Exception

File: src/main/SimpleListImpl.java [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+            arr = new Object[temp.length + 1];
+            System.arraycopy(temp, 0, arr, 0, index);
+            arr[index] = element;
+            System.arraycopy(temp, index, arr, index + 1, temp.length-index);
+        }
+    }
+
+    @Override
+    public Object remove(int index) {
+        if (index < size()) {
```
Comment (drykiss0) [inline src/main/SimpleListImpl.java]: Please handle when index >= size()

File: src/main/SimpleListImpl.java [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+        return arr.length;
+    }
+
+    @Override
+    public boolean isEmpty() {
+        if (size() == 0) {
+            return true;
+        } else {
+            return false;
+        }
```
Comment (drykiss0) [inline src/main/SimpleListImpl.java]: Please simplify this expression.

File: src/main/SimpleListImpl.java [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+        System.arraycopy(temp, 0, arr, 0, temp.length);
+        arr[arr.length - 1] = o;
+        return true;
+    }
+
+    @Override
+    public boolean remove(Object o) {
+        int j = 0;
+        for (int i = 0; i < size(); i++) {
+            if (o == arr[i]) {
```
Comment (drykiss0) [inline src/main/SimpleListImpl.java]: Please use **equals** for comparisons.

File: src/main/SimpleListImpl.java [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+                int aftterI = temp.length - i - 1;
+                System.arraycopy(temp, i + 1, arr, i, aftterI);
+                j++;
+            }
+        }
+        if (j > 0) {
+            return true;
+        } else {
+            throw new UnsupportedOperationException();
+        }
```
Comment (drykiss0) [inline src/main/SimpleListImpl.java]: Why do you throw Exception if an Element is not found? Please fix.

File: src/main/SimpleListImpl.java [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    @Override
+    public void clear() {
+        if (arr.length == 0) {
+            throw new UnsupportedOperationException();
```
Comment (drykiss0) [inline src/main/SimpleListImpl.java]: You should be able to clear empty array as well. Why do you throw Exception?

File: src/main/SimpleListImpl.java [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+        } else {
+            arr = new Object[0];
+        }
+    }
+
+    @Override
+    public boolean contains(Object o) {
+        for (int i = 0; i < size(); i++) {
+            if (arr[i] == o) {
```
Comment (drykiss0) [inline src/main/SimpleListImpl.java]: Please use **equals** for comparison.

File: src/main/SimpleListImpl.java [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+                return true;
+            } else {
+            }
```
Comment (drykiss0) [inline src/main/SimpleListImpl.java:119]: Please clean up

**Review by drykiss0 — CHANGES_REQUESTED** [review]
Please address issues.

File: src/main/SimpleListImpl.java (hunk 8 above)
Comment (OmelchenkoE) [reply src/main/SimpleListImpl.java]: then will exception

Comment (drykiss0) [reply src/main/SimpleListImpl.java]: > then will exception

You're right, sorry. I meant - you need to handle situation when index < 0

**Pull request merged**